firmware/lib/Axis/src/Axis.cpp

```diff
@@ -0,0 +1,317 @@
+#include "Axis.h"
+#include "Debug.h"
+#include "Timer.h"
+
+// -------------------------------- X AXIS SETUP -------------------------------- //
+#define ACCEL_X                 100
+#define VEL_MIN_X               100
+#define VEL_MAX_X               1000
+#define STEP_PIN_X              5 // need to know the register for fast operation pin 5 - PC25 
+#define DIR_PIN_X               6 // need to know the register for fast operation pin 6 - PC24
+#define ENCODER_PIN_A_X         7
+#define ENCODER_PIN_B_X         8
+#define LIMIT_SW_HOME_PIN_X     9
+#define LIMIT_SW_FAR_PIN_X      10
+
+AxisConfig axis_x_config = {DIR_PIN_X, STEP_PIN_X, ENCODER_PIN_A_X, ENCODER_PIN_B_X, LIMIT_SW_HOME_PIN_X, LIMIT_SW_FAR_PIN_X,
+                            ACCEL_X, VEL_MAX_X, VEL_MIN_X, TC1, 0, 1, TC3_IRQn, TC4_IRQn, isr_encoder_x, isr_limit_switch_x};
+
+Axis setup_axis(AxisConfig *axis_config)
+{  
+    // Set up the timer interrupt
+    // TC1 channel 0, the IRQ for that channel and the desired frequency
+    pinMode(axis_config->encoder_pin_a, INPUT);
+    pinMode(axis_config->encoder_pin_b, INPUT);
+    Encoder encoder_x = {axis_config->encoder_pin_a, 10, 0};
+
+    pinMode(axis_config->ls_home, INPUT_PULLUP);
+
+    LimitSwitch ls_home = {axis_config->ls_home, digitalRead(axis_config->ls_home)};
+
+    pinMode(axis_config->ls_far, INPUT_PULLUP);
+    LimitSwitch ls_far_from_home = {axis_config->ls_far, digitalRead(axis_config->ls_far)};
+    
+    pinMode(axis_config->dir_pin, OUTPUT);
+    pinMode(axis_config->step_pin, OUTPUT);
+    Axis axis_x_temp = {axis_config->dir_pin, axis_config->step_pin, axis_config->acceleration, axis_config->vel_max, axis_config->vel_min, 
+                        {0, 0, 0}, ACCELERATE, encoder_x, ls_home, ls_far_from_home,
+                        axis_config->timer, axis_config->channel_velocity, axis_config->channel_velocity, axis_config->isr_velocity, axis_config->isr_accel};
+
+    // attach interrupts
+    attachInterrupt(digitalPinToInterrupt(axis_config->encoder_pin_a), axis_config->isr_encoder, RISING);
+    attachInterrupt(digitalPinToInterrupt(axis_config->ls_home), axis_config->isr_limit_switch, CHANGE);
+    attachInterrupt(digitalPinToInterrupt(axis_config->ls_far), axis_config->isr_limit_switch, CHANGE);
+    return axis_x_temp;
+}
+
+Axis axis_x = setup_axis(&axis_x_config);
```

Comment:
Don't love the idea of calling functions just in the global scope outside of any other function.
Especially ones that have side effects like setting pinModes and attaching interrupts.

Can we move these to some init function that can be called from main or just call them from main?